pkg/controller/conditions/conditions.go

```diff
@@ -0,0 +1,343 @@
+	for _, d := range dependents {
+		if d == happy || slices.Contains(deps, d) {
+			continue
+		}
+		deps = append(deps, d)
+	}
+
+	m := Manager{
+		accessor:   accessor,
+		happy:      happy,
+		dependents: deps,
+	}
+
+	m.initializeConditions()
+
+	return &m
+}
+
+// initializeConditions ensures that the conditions for the manager and its dependents are properly
+// initialized. Specifically, it initializes the "happy" condition and sets the initial status for
+// each dependent condition.
+//
+// The method performs the following:
+//  1. Retrieves the "happy" condition using. If it does not exist, it creates a new one with
+//     `ConditionUnknown` status and sets it.
+//  2. Sets the status of each dependent condition based on the "happy" condition's status. If the
+//     "happy" condition's status is `True`, all dependent conditions are set to `True`, otherwise
+//     to `Unknown`.
+func (r *Manager) initializeConditions() {
+	happy := r.GetCondition(r.happy)
+	if happy == nil {
+		happy = &common.Condition{
+			Type:   r.happy,
+			Status: metav1.ConditionUnknown,
+		}
+		r.SetCondition(*happy)
+	}
+
+	status := metav1.ConditionUnknown
+	if happy.Status == metav1.ConditionTrue {
+		status = metav1.ConditionTrue
+	}
+
+	for _, t := range r.dependents {
+		if c := r.GetCondition(t); c != nil {
+			continue
+		}
+
+		r.SetCondition(common.Condition{
+			Type:   t,
+			Status: status,
+		})
+	}
+}
+
+func (r *Manager) IsHappy() bool {
+	if r.accessor == nil {
+		return false
+	}
+
+	return IsStatusConditionTrue(r.accessor.GetConditions(), r.happy)
+}
+
+func (r *Manager) GetTopLevelCondition() *common.Condition {
+	return r.GetCondition(r.happy)
+}
+
+func (r *Manager) GetCondition(t string) *common.Condition {
+	return FindStatusCondition(r.accessor.GetConditions(), t)
+}
+
+// SetCondition sets the given condition on the manager. It updates the list of conditions and
+// sorts them alphabetically. After updating, it recomputes the happiness state based on the
+// provided condition type.
+//
+// Parameters:
+//   - `cond`: The condition to set. The condition will be added or updated based on its type.
+func (r *Manager) SetCondition(cond common.Condition) {
+	if r.accessor == nil {
+		return
+	}
+
+	conditions := r.accessor.GetConditions()
+
+	if !SetStatusCondition(&conditions, cond) {
+		return
+	}
+
+	r.accessor.SetConditions(conditions)
+	r.RecomputeHappiness(cond.Type)
+}
+
+// ClearCondition removes the specified condition type from the manager's list of conditions
+// and recomputes happiness.
+//
+// Parameters:
+//   - `t`: The type of the condition to remove.
+//
+// Returns:
+//   - `nil` if the condition was removed successfully or was not found.
+func (r *Manager) ClearCondition(t string) error {
+	if r.accessor == nil {
+		return nil
+	}
+
+	conditions := r.accessor.GetConditions()
+
+	if !RemoveStatusCondition(&conditions, t) {
+		return nil
+	}
+
+	slices.SortFunc(conditions, func(a, b common.Condition) int {
```

Comment:
No sorting in SetCondition?
But with sorting probably [When designing Conditions for a resource, it's helpful to have a common top-level condition which summarizes more detailed conditions. Simple consumers may simply query the top-level condition. Although they are not a consistent standard, the Ready and Succeeded condition types may be used by API designers for long-running and bounded-execution objects, respectively.](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) is not applicable?

Reply:
> No sorting in SetCondition?

the sort here is probably a leftover of various iterations, I don't think it does make any sense here (probably also the sort done in the `SetCondition` is probably not needed)